design/StableRegions-20230209.md

```diff
@@ -0,0 +1,222 @@
+     table is suitable to do a "batch" rebuild of the dynamic heap-allocated vectors
+     in that table, which get lost during upgrades.
+
+ - 32768 entries (statically sized).
+ - 4 bytes per entry.
+ - entry type = `BlockRegion { region : Nat16, position : Nat16 }`
+ - the location of each entry gives its corresponding block ID.
+
+### region-blocks table
+
+ - purpose: 
+   - relate a region ID to its vector of block IDs, to compute an access efficiently (load or store).
+
+   - NB: The organization of this table is meant to support O(1) load
+     and store operations, but in so doing, it needs to use a
+     dynamically-sized vector for each region.  There is no a priori
+     way to know how to allocate these, and if we naively preallocate
+     them to each be the potential "maximal region" (with all blocks
+     allocated to it), the resulting preallocated table requires
+     gigabytes of space, and thus is prohibitively large.
+
+     At the same time, having dynamically-sized stable vectors is kind
+     of the point of regions (each is such a vector, in a sense), and
+     so requiring these first to implement regions' meta data seems
+     potentially circular in concept, if not also very complex.
+
+     So, we seem forced to use dynamically-sized heap vectors, and to
+     deal the compilations of integrating them with GC (how involved is
+     that?) and rebuilding them on upgrade.
+
+
+ - 32768 entries (statically sized, ignoring the `vec_ptr` objects referenced therein).
+ - 16 bytes per entry.
+ - entry type = `RegionBlocks { size_in_pages: Nat64; vec_capacity : Nat32; vec_ptr: Nat32 }`
+ - the location of each entry in the table gives its corresponding region ID.
+ - the `size_in_pages` field measures the size of the region in terms of _pages_, not blocks. 
+ - the `vec_capacity` and `vec_ptr` fields work with `size_in_pages`
+   to represent a growable vector that we call the region's **"access
+   vector"** (because "blocks vector" sounds a bit strange, and its
+   used to support O(1) access operations): 
+   - the access vector's address is held in `vec_ptr` and it has `vec_capacity` slots.
+   - the first `size_in_pages / 128` slots of `vec_ptr` contain a valid page block ID for the region.
+   - to support dynamic growth, _the access vector is held in the Motoko **heap** rather than **stable memory**._
+   - the access vector doubles when it grows.
+   - no region has more than 32k page blocks, so a `Nat16` suffices for `capacity`,
+   - but we use a `Nat32` for `capacity` to make all fields things word-aligned (does that matter?).
+   - during an upgrade, the access vectors are "rebuilt" in a batch (see `rebuild`).
+
+
+### Overview of `rebuild`
+
+When a canister upgrades, its heap is lost and the access vectors for each region are lost too (see `region-blocks` table section).
+
+We use the `block-region` and `region-blocks` tables in stable memory to rebuild these vectors in the latter table:
+
```

Comment:
The `region-blocks` table would be only stored in heap, isn't it? The `block-region` would then allow restoring the `region-blocks` tables? I think the size of the region can be also inferred by scanning the `block-regions` and reconstructing the `region-blocks`, e.g. if we use `Buffer` and simply add all page block addresses of that region in any order (random-access and lazy extension of buffer size).

Reply:
I'm confused why we need a region-blocks table in the heap at all and can't just store the entries of  the table as heap objects, the region values themselves. It's just an extra indirection that will need to be chased for every read/write.
That does mean storing the vectors and  page counts directly in the region heap objects and treating them as new sorts of object tags, but that seems doable. When we grow a region, we update the stable memory data accordingly to keep it in sync.
Every load/store can then consult the fields of the region object to do the logical 2 physical address translations necessary.

Reply:
> The region-blocks table would be only stored in heap, isn't it? 

> I'm confused why we need a region-blocks table in the heap at all

Except for the access vectors, the `region-blocks` table is stored entirely in stable memory, not the heap.  

Why have both?

In addition to heap pointers for the access vectors, it contains the sizes of each region, which the `block-region` table lacks.  

>  I think the size of the region can be also inferred by scanning the block-regions and reconstructing the region-blocks,

Yes, we could reconstruct these sizes somewhat imprecisely, but not precisely enough to recover the original page count.

(Recall that the size of a region is measured in **pages** (not blocks), and that each block is 128 pages So, counting blocks is not quite precise enough to recover the actual page count.)

Note that we store `pages` as `Nat64` for each region, just like the Rust version of this code.